File: proj/OC/oneOC_mac/Ctrls/JoinCtrl.m
//
//  JoinCtrl.m
//  oneOC_mac
//
//  Created by Jz D on 2020/4/2.
//  Copyright © 2020 Jz D. All rights reserved.
//

#import "JoinCtrl.h"

#include "GCDAsyncSocket.h"


#import "PacketH.h"


@interface JoinCtrl ()<NSNetServiceDelegate, NSNetServiceBrowserDelegate, GCDAsyncSocketDelegate, NSTableViewDelegate, NSTableViewDataSource>
 
@property (strong, nonatomic) GCDAsyncSocket *socket;
@property (strong, nonatomic) NSMutableArray *services;
@property (strong, nonatomic) NSNetServiceBrowser *serviceBrowser;



@property (weak) IBOutlet NSTableView *tableView;



@end

@implementation JoinCtrl


static NSString *ServiceCell = @"ServiceCell";


- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"加入者";

    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    // Start Browsing
    [self startBrowsing];
    
    
    [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(didSelectRow:) name: NSTableViewSelectionDidChangeNotification object: self.tableView];
    
}



- (void)didSelectRow: (NSNotification * ) noti{
    NSTableView * tb = (NSTableView *) noti.object;
    NSInteger row = tb.selectedRow;
    // Fetch Service
    NSNetService *service = [self.services objectAtIndex: row];

    // Resolve Service
    NSLog(@"Resolve 一下， 解析处理");
    service.delegate = self;
    // 点击，服务就 gg
   
    [service resolveWithTimeout:30.0];

    
}






- (void)viewWillDisappear{
    [super viewWillDisappear];

    // Stop Browsing Services
    [self stopBrowsing];
 
}





- (void)cancel:(id)sender {
       // Notify Delegate
       [self.delegate controllerDidCancelJoining:self];
    
       // Stop Browsing Services
       [self stopBrowsing];
    
       // Dismiss View Controller
       [self dismiss];
}




- (void)dealloc {
    if (_delegate) {
        _delegate = nil;
    }


    [[NSNotificationCenter defaultCenter] removeObserver: self];
}


- (void)startBrowsing {
    if (self.services) {
        [self.services removeAllObjects];
    } else {
        self.services = [[NSMutableArray alloc] init];
    }
 
    // Initialize Service Browser
    self.serviceBrowser = [[NSNetServiceBrowser alloc] init];
 
    // Configure Service Browser
    [self.serviceBrowser setDelegate:self];
    [self.serviceBrowser searchForServicesOfType:@"deng._tcp." inDomain:@"local."];
}



- (void)stopBrowsing {
    if (self.serviceBrowser) {
        [self.serviceBrowser stop];
        [self.serviceBrowser setDelegate:nil];
        [self setServiceBrowser:nil];
    }
}



#pragma mark - Table view data source




- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView{
    if (self.services){
        return [self.services count];
    }else{
        return 0;
    }
}



- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row{
    NSTextField *cell = [[NSTextField alloc] init];
    
     
    // Fetch Service
    NSNetService *service = [self.services objectAtIndex: row];
    
    cell.stringValue = service.name;
    
    return cell;
}






- (void)netServiceBrowser:(NSNetServiceBrowser *)serviceBrowser didFindService:(NSNetService *)service moreComing:(BOOL)moreComing {
    // Update Services
    [self.services addObject:service];
 
    if(!moreComing) {
        // Sort Services
        [self.services sortUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES]]];
 
        // Update Table View
        [self.tableView reloadData];
    }
}


- (void)netServiceBrowser:(NSNetServiceBrowser *)serviceBrowser didRemoveService:(NSNetService *)service moreComing:(BOOL)moreComing {
    // Update Services
    [self.services removeObject:service];
 
    if(!moreComing) {
        // Update Table View
        [self.tableView reloadData];
    }
}


- (void)netServiceBrowserDidStopSearch:(NSNetServiceBrowser *)serviceBrowser {
    [self stopBrowsing];
}



- (void)netServiceBrowser:(NSNetServiceBrowser *)aBrowser didNotSearch:(NSDictionary *)userInfo {
    [self stopBrowsing];
}



- (void)netServiceDidResolveAddress:(NSNetService *)service {
    // Connect With Service
    NSLog(@"Deng: Connect With Service");
    if ([self connectWithService:service]) {
        NSLog(@"Did Connect with Service: domain(%@) type(%@) name(%@) port(%i)", [service domain], [service type], [service name], (int)[service port]);
    } else {
        NSLog(@"XXX: Unable to Connect with Service: domain(%@) type(%@) name(%@) port(%i)", [service domain], [service type], [service name], (int)[service port]);
    }
}



- (BOOL)connectWithService:(NSNetService *)service {
    BOOL _isConnected = NO;
 
    // Copy Service Addresses
    NSArray *addresses = [[service addresses] mutableCopy];
 
    if (!self.socket || ![self.socket isConnected]) {
        // Initialize Socket
        NSLog(@"Initialize Socket, 新建了 Socket");
        self.socket = [[GCDAsyncSocket alloc] initWithDelegate:self delegateQueue:dispatch_get_main_queue()];
 
        // Connect
        while (!_isConnected && [addresses count]) {
            NSData *address = [addresses objectAtIndex:0];
 
            NSError *error = nil;
            if ([self.socket connectToAddress:address error:&error]) {
                _isConnected = YES;
 
            } else if (error) {
                NSLog(@"Unable to connect to address. Error %@ with user info %@.", error, [error userInfo]);
            }
        }
 
    } else {
        _isConnected = [self.socket isConnected];
    }
 
    return _isConnected;
}






// 读数据
- (void)socket:(GCDAsyncSocket *)socket didConnectToHost:(NSString *)host port:(UInt16)port {
    NSLog(@"Socket Did Connect to Host: %@ Port: %hu", host, port);
    NSLog(@"邓: 故事发生一");

    
       // Notify Delegate
       [self.delegate controller:self didJoinGameOnSocket:socket];
    
       // Stop Browsing
       [self stopBrowsing];
    
       // Dismiss View Controller
       [self dismiss];
}




- (void) dismiss{
    // Dismiss View Controller
    if (self.presentingViewController) {
        [self.presentingViewController dismissViewController:self];
    } else {
        //for the 'show' transition
        [self.view.window close];
    }
}







@end






/*
    Browsing Services
 
 
 
    To browse for services on the local network, we use the NSNetServiceBrowser class. Before putting the NSNetServiceBrowser class to use, we need to create a few private properties. Add a class extension to the MTJoinGameViewController class and declare three properties as shown below.
 
 
 
 The first property, socket of type GCDAsyncSocket, will store a reference to the socket that will be created when a network service resolves successfully.
 
 
 
 
 2
 
 The services property (NSMutableArray) will store all the services that the service browser discovers on the network. Every time the service browser finds a new service, it will notify us and we can add it to that mutable array. This array will also serve as the data source of the view controller's table view.


 The third property, serviceBrowser, is of type NSNetServiceBrowser and will search the network for network services that are of interest to us. Also note that the MTJoinGameViewController conforms to three protocols. This will become clear when we implement the methods of each of these protocols.
 
 */

File: project/socketG/socketG/Ctrls/JoinCtrl.swift
//
//  JoinCtrl.swift
//  socketG
//
//  Created by Jz D on 2020/4/26.
//  Copyright © 2020 Jz D. All rights reserved.
//

import Cocoa


protocol JoinListCtrlDelegate: class{

    func didJoinTask(c controller: JoinCtrl, on socket: GCDAsyncSocket)
}




class JoinCtrl: NSViewController {
    
    var services = [NetService]()
    var socket: GCDAsyncSocket?
    var serviceBrowser: NetServiceBrowser?
    
    weak var delegate: JoinListCtrlDelegate?
    
    
    
    @IBOutlet weak var tableView: NSTableView!
    
    

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "加入者"

        tableView.delegate = self
        tableView.dataSource = self
        
        // Start Browsing
        startBrowsing()
        NotificationCenter.default.addObserver(self, selector: #selector(didSelect(row:)), name: NSTableView.selectionDidChangeNotification, object: tableView)
        
    }
    
    

    override func viewWillDisappear() {
        super.viewWillDisappear()
        cancel()
    }
    
    

    @objc
    func didSelect(row noti: NSNotification){
        if let tb = noti.object as? NSTableView{
            let row = tb.selectedRow
            
            // Fetch Service
            let service = services[row]
            service.delegate = self;
            // 点击，服务就 gg
            service.resolve(withTimeout: 30.0)
        }
        
    }
    
    
      func cancel(){
          stopBrowsing()
      }
      
      

      func startBrowsing(){
          services = []
       
          // Initialize Service Browser
          serviceBrowser = NetServiceBrowser()
       
          // Configure Service Browser
          serviceBrowser?.delegate = self
          serviceBrowser?.searchForServices(ofType: "_deng._tcp.", inDomain: "local.")
      }

      
    
    func stopBrowsing(){
        serviceBrowser?.stop()
        serviceBrowser?.delegate = nil
        serviceBrowser = nil
    }

    
    
}




extension JoinCtrl: NetServiceDelegate, NetServiceBrowserDelegate{
    
    
    
    func connectWith(service s: NetService) -> Bool{
        var isConnected = false
     
        // Copy Service Addresses
        guard let addresses = s.addresses else{
            return false
        }
        
        var condition = false
        
        if let ss = socket{
            if ss.isConnected == false{
                condition = true
            }
            else{
                isConnected = ss.isConnected
            }
        }
        if socket == nil || condition{
            
            // Initialize Socket
            
            socket = GCDAsyncSocket(delegate: self, delegateQueue: DispatchQueue.main)
            
            // Connect
            while isConnected == false, addresses.count > 0 {
                let address = addresses[0]
                
                do {
                    if let sss = socket{
                        try sss.connect(toAddress: address)
                        // 结果 bool ,
                        //  就是 ok,
                        //  不 ok, 顺带 error 信息
                        isConnected = true
                    }
                } catch {
                    print("Unable to connect to address. Error \(error) with user info ")
                }
            }
        }
     
        return isConnected;
    }



    
    
    
    func netServiceBrowser(_ browser: NetServiceBrowser, didFind service: NetService, moreComing: Bool) {
        
        
        // Update Services
        services.append(service)
     
        if !moreComing{
            // Sort Services
            services.sort { (lhs, rhs) -> Bool in
                lhs.name > rhs.name
            }
            // Update Table View
            tableView.reloadData()
        }
    }


    
    func netServiceBrowser(_ browser: NetServiceBrowser, didRemove service: NetService, moreComing: Bool) {
        // Update Services
        if let index = services.firstIndex(where: {  (s) -> Bool in
            s == service
        }){
             services.remove(at: index)
        }
      
        if !moreComing{
            // Update Table View
            tableView.reloadData()
        }
    }


    func netServiceBrowserDidStopSearch(_ browser: NetServiceBrowser) {
        stopBrowsing()
        
    }



    func netServiceBrowser(_ browser: NetServiceBrowser, didNotSearch errorDict: [String : NSNumber]) {
        stopBrowsing()
        
    }

    
    func netService(_ sender: NetService, didNotResolve errorDict: [String : NSNumber]) {
        sender.delegate = nil
    }


    func netServiceDidResolveAddress(_ sender: NetService) {

        // Connect With Service
     
        if connectWith(service: sender){
            print("Did Connect with Service:  domain(\(sender.domain)) type(\(sender.type)) name(\(sender.name)) port(\(sender.port)")
        }
        else{
            print("Unable to Connect with Service:  domain(\(sender.domain)) type(\(sender.type)) name(\(sender.name)) port(\(sender.port)")
        }
    }


  
}


extension JoinCtrl: GCDAsyncSocketDelegate{
    
    
    func socket(_ sock: GCDAsyncSocket, didConnectToHost host: String, port: UInt16) {

           print("Socket Did Connect to Host: \(host) Port: \(port)")
        
           // Notify Delegate
           delegate?.didJoinTask(c: self, on: sock)

           // Stop Browsing
           stopBrowsing()
        
           // Dismiss View Controller
           dismiss()
    }



}




extension JoinCtrl: NSTableViewDataSource{
    func numberOfRows(in tableView: NSTableView) -> Int {
        return services.count
    }
    
    
}



extension JoinCtrl: NSTableViewDelegate{
    func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        let cell = NSTextField()
        cell.stringValue = services[row].name
        return cell
    }
    
    
}



